compiler/src/iree/compiler/Codegen/LLVMCPU/test/set_transform_strategy.mlir

```diff
@@ -0,0 +1,64 @@
+
+#pipeline_layout = #hal.pipeline.layout<push_constants = 0, sets = [
+  #hal.descriptor_set.layout<0, bindings = [
+    #hal.descriptor_set.binding<0, storage_buffer>,
+    #hal.descriptor_set.binding<1, storage_buffer>
+  ]>
+]>
+hal.executable private @pack  {
+  hal.executable.variant public @embedded_elf_x86_64, target = <"llvm-cpu", "embedded-elf-x86_64", {
+    cpu_features = "+avx512f,+avx2",
+    data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128",
+    native_vector_size = 16 : index,
+    target_triple = "x86_64-unknown-unknown-eabi-elf"
+  }> {
+    hal.executable.export public @pack layout(#pipeline_layout) {
+    ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index):
+      %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg1, %arg2
+      hal.return %x, %y, %z : index, index, index
+    }
+    builtin.module {
+      func.func @pack() {
+        %c0 = arith.constant 0 : index
+        %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<readonly:tensor<384x512xf32>>
+        %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<24x512x16x1xf32>>
+        %2 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [384, 512], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<384x512xf32>> -> tensor<384x512xf32>
+        %3 = tensor.empty() : tensor<24x512x16x1xf32>
+        %pack = tensor.pack %2 inner_dims_pos = [0, 1] inner_tiles = [16, 1] into %3: tensor<384x512xf32> -> tensor<24x512x16x1xf32>
+        flow.dispatch.tensor.store %pack, %1, offsets = [0, 0, 0, 0], sizes = [24, 512, 16, 1], strides = [1, 1, 1, 1] : tensor<24x512x16x1xf32> -> !flow.dispatch.tensor<writeonly:tensor<24x512x16x1xf32>>
+        return
+      }
+    }
+  }
+}
+// CHECK-DAG:   #[[TRANSLATION:.+]] = #iree_codegen.translation_info<TransformDialectCodegen>
+// CHECK:         hal.executable.export public @pack
+// CHECK-SAME:       translation_info = #[[TRANSLATION]]
+// CHECK-LABEL: func.func @pack
+// CHECK:         transform.sequence
+// CHECK:         %[[PACK:.+]] = transform.structured.match ops{["tensor.pack"]} in %arg0
+// CHECK:         %[[FORALL:.+]], %[[TILED_PACK:.+]] = transform.iree.tile_to_forall_and_workgroup_count_region
+// CHECK-SAME:      %[[PACK:.+]] num_threads [1]
+// CHECK:         %[[CAST:.+]] = cast %[[TILED_PACK]] : !pdl.operation to !transform.op<"tensor.pack">
+// CHECK:         %[[PAD:.+]], %[[RESHAPE:.+]], %[[TRANS:.+]] =
+// CHECK-SAME:      transform.structured.lower_pack %[[CAST]]
+// CHECK:         %[[TILED_TRANS:.+]], %{{.+}} = transform.structured.tile %[[TRANS]][1, 8, 8]
```

Comment:
The AVX512 would be very interesting. We could close the gap between codegen and ukernels if we have it implemented. There are 5% gap in `pack_LHS(384x512) -> 24x512x16x1` case and 15% gap in `pack_LHS(384x384) -> 24x512x16x1`. The codegen [ASM dump](https://gist.githubusercontent.com/hanhanW/d1e3eee0bb952c1654a66349b24c197e/raw) are all about ymm registers, while ukernels work on zmms. I believe that we'll see zmm registers if we have avx512 implementation. I file https://github.com/openxla/iree/issues/13127 for tracking it.

Reply:
Can you share a pointer to the ukernel? We can check if it's the same approach as the one in the paper. Note that wider vectors are not necessarily faster in a transpose op. The larger the number of elements is, the larger the dependence chain is (IOW, transposing more elements at once means more instructions needed to place each element in its output position, and more cycles from the consumers of those elements waiting for them). In this regard, the AVX512 approach uses only one 512-bit register as a "virtual" 2D vector registers to perform the transposition (i.e., only 4x4 elements transposed at a time), which should be better than the 4x8 and 8x8 configurations that we have for AVX2.

> Note that this may require going to "higher dimensional transposes".

Not sure what this means exactly but I added support for n-D transposes for the AVX2 patterns. They are decomposed and treated as 2D transposes and some of them become plain memcopies (without transposition) so we should be getting all of that working out of the box.

Reply:
> Can you share a pointer to the ukernel?

Here is the entry point: https://sourcegraph.com/github.com/openxla/iree@d474394200ee504c394818b9661db5656c2472b5/-/blob/runtime/src/iree/builtins/ukernel/arch/x86_64/pack_x86_64_avx512_base.c?L54-66

I also attach a [dump](https://gist.githubusercontent.com/hanhanW/29aca38f0479be6e3d5dc1bdd945690a/raw) of ukernels from `perf annotate`. It includes source and asm code.